src/python/pants/backend/go/util_rules/build_pkg.py

```diff
@@ -226,6 +238,36 @@ class GoCompileActionIdResult:
     action_id: str
 
 
+# TODO: Merge this rule helper and the AssemblyPostCompilationRequest.
+@rule_helper
+async def _add_objects_to_archive(
+    input_digest: Digest,
+    pkg_archive_path: str,
+    obj_file_paths: Iterable[str],
+) -> ProcessResult:
+    # Use `go tool asm` tool ID since `go tool pack` does not have a version argument.
+    asm_tool_id = await Get(GoSdkToolIDResult, GoSdkToolIDRequest("asm"))
+    pack_result = await Get(
+        ProcessResult,
+        GoSdkProcess(
+            input_digest=input_digest,
+            command=(
+                "tool",
+                "pack",
+                "r",
+                pkg_archive_path,
+                *obj_file_paths,
+            ),
+            env={
+                "__PANTS_GO_ASM_TOOL_ID": asm_tool_id.tool_id,
```

Comment:
Presumably this is to invalidate when the go toolchain is upgraded?

Reply:
Yes. Adding such environment variables is necessary to prevent caching issues with Go toolchain upgrades that we've already witnessed previously in practice.